include/pipe.cc

```diff
@@ -1,26 +1,83 @@
-// Constructor with id and node pointers
+// Constructor with id, node pointers, diameter and status
 pipenetwork::Pipe::Pipe(
-    unsigned id, const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes)
-    : id_{id}, nodes_{nodes} {
+    unsigned id, const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes,
+    const double diameter, bool status)
+    : id_{id}, nodes_{nodes}, radius_{diameter / 2}, isopen_{status} {
   length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
 }
 
-// Calculate and return discharge using Darcy-Weisbach equation
-double pipenetwork::Pipe::discharge() {
+// Constructor with id, node pointers, diameter, status and max allowable
+// velocity
+pipenetwork::Pipe::Pipe(
+    unsigned id, const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes,
+    const double diameter, bool status, const double max_velocity)
+    : id_{id},
+      nodes_{nodes},
+      radius_{diameter / 2},
+      isopen_{status},
+      max_velocity_{max_velocity} {
+  length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
+}
+
+// Calculate discharge using Darcy-Weisbach equation
+//! dhead/length =
+//! (8*darcy_factor*pow(discharge,2)/(pow(M_PI,2)*g*pow(2*radius,5)) That is,
+//! discharge =
+//! sqrt(dhead/length*pow(M_PI,2)*g*pow(2*radius,5)/(8*darcy_friction));
+void pipenetwork::Pipe::compute_discharge_darcy_weisbach() {
   // The Darcy-Weiabach equation is only applicable when heads at both ends of
   // the pipe are known, thus check it
   if (nodes_.at(0)->ishead() && nodes_.at(1)->ishead()) {
     const double dhead = nodes_.at(0)->head() - nodes_.at(1)->head();
     discharge_ = sqrt(std::abs(dhead) * pow(M_PI, 2) * pipenetwork::Gravity(2) *
-                      pow(2 * radius_, 5) / (8 * darcy_friction_));
+                      pow(2 * radius_, 5) / (8. * length_ * darcy_friction_));
     // defined flow direction from nodes_.at(0) to nodes.at(1) as positive
     if (dhead < 0) discharge_ *= -1.;
   } else {
     throw std::runtime_error(
         "Unknown head exists, cannot calculate discharge using Darcy Weisbach "
         "equation");
   }
-  return discharge_;
+}
+
+// Calculate discharge using Hazen-Williams equation
+//! dhead/length =
+//! (10.67*pow(discharge,1.852)/(pow(pipe_roughness,1.852)*pow(2*radius,4.8704))
```

Comment:
Does these variables 10.67, 1.852 depend on a certain unit? How do we ensure our input is consistent with that?

Reply:
10.67 depends on the certain unit (m and s) 1.852 not.
Both the Hazen-Williams and Darcy-Weisbach equation I used take m and s as unit of input parameter, I think we can convert the units to these when reading the file?